query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -106,37 +106,37 @@ impl<F> FnArgInto<F> for FnArg where F: Fn(Variable) -> Result<TypedValue> {
 /// ignore.  It ensures that each row has the same length, and drops the columns that have `false`
 /// in the `column_is_bound` sequence.  Finally, it ensures that each column is homogeneously typed,
 /// and returns a sequence of the derived type for each column.
-fn reduced_rows_and_column_types(rows: Vec<Vec<TypedValue>>, column_is_bound: &Vec<bool>) -> Result<(Vec<Vec<TypedValue>>, Vec<ValueType>)> {
-    let mut matrix = Vec::with_capacity(rows.len());
+fn reduced_rows_and_column_types(rows: Vec<Vec<TypedValue>>, column_is_bound: &Vec<bool>) -> Result<(Vec<TypedValue>, Vec<ValueType>)> {
+    let expected_width = column_is_bound.iter().filter(|x| **x).count();
+    let mut matrix = Vec::with_capacity(rows.len() * expected_width);
+
+    if rows.is_empty() {
+        bail!(ErrorKind::InvalidGroundConstant)
+    }
 
     for row in rows {
         if column_is_bound.len() != row.len() {
             bail!(ErrorKind::InvalidGroundConstant)
         }
 
-        let mut retained_row = Vec::with_capacity(column_is_bound.len());
         for (is_bound, value) in column_is_bound.into_iter().zip(row.into_iter()) {
             if *is_bound {
-                retained_row.push(value);
+                matrix.push(value);
             }
         }
-
-        matrix.push(retained_row);
     }
 
-    let column_types: Vec<ValueType> = match matrix.get(0) {
-        None => {
-            bail!(ErrorKind::InvalidGroundConstant)
-        },
-        Some(first_row) => {
-            first_row.iter().map(|x| x.value_type()).collect()
-        },
-    };
+    let column_types: Vec<ValueType>;
+
+    {     // Scoped borrow of matrix.
+        let first_row = &matrix[0..expected_width];
+        column_types = first_row.iter().map(|x| x.value_type()).collect();
 
-    for row in matrix.iter() {
-        for (column_type, value) in column_types.iter().zip(row.into_iter()) {
-            if *column_type != value.value_type() {
-                bail!(ErrorKind::InvalidGroundConstant)
+        for row in matrix.chunks(expected_width) {
```

Comment:
What does this do when `expected_width` is 0?  I think that's possible, and this might fail.  A needed test case, I think.